py-data-manipulation.Rmd

```diff
@@ -0,0 +1,457 @@
+dog_licenses = dog_licenses.drop('new_column', axis = 1)
+```
+
+We can also create new columns by doing operations on existing columns.
+For example, let's create a column with the length of the dog's name.
+
+```python
+# the method .str allows string functions to be applied to the column contents (here we use 'len()')
+name_lengths = dog_licenses['animal_name'].str.len()
+
+# create a new column called 'name_length'
+dog_licenses['name_length'] = name_lengths
+
+# we can see that the new column has been added to the end of the data frame
+dog_licenses.columns
+```
+
+Let's create a column with the age of each animal at the time it was licensed.
+Note: this step takes a few seconds to run.
+
+```python
+# use to_datetime to create a datetime object for each date we're using
+animal_birth_date = pd.to_datetime(dog_licenses['animal_birth_month'])
+license_issued_datetime = pd.to_datetime(dog_licenses['license_issued_date'])
+
+# subtract the birth date from the license date
+age_when_licensed = license_issued_datetime - animal_birth_date
+
+# assign age_when_licensed to a new column
+dog_licenses['age_when_licensed'] = age_when_licensed
+```
+
+## How can I tell what's gone wronge in my programs?
+
+Typos and errors are inevitable - they happen to everyone. Python provides
+error messages if it can't run your code that will help you figure out where
+the problem is.
+
+```
+# this code generates a syntax error
+dog_licenses['animal_name'
+```
+
+The very bottom of the error message is the most helpul place to start looking:
+this says what kind of error it is (e.g. `SyntaxError`, `ValueError`,
+`TypeError`, etc) and sometimes includes a message that describes what went
+wrong.
+
+It also points to the place where the error happened: in this case, there's a
+missing square bracket, so the error we got was that Python reached the end of
+the file before it expected to because we didn't include proper punctuation.
+
+```python
+# this code generates a runtime error (there's nothing wrong with the syntax,
+# but Python can't execute it because of a different error)
+
+dog_licenses['animal_name'] < dog_licenses['zip_code']
+```
+
+The last line again tells us what went wrong, but this time we get a
+*traceback*: Python is showing us the series of steps that it followed
+internally before it hit an error. This is helpful because often the error will
+ultimately occur inside a function we're using but whose code we haven't ever
+seen, so Python shows us both what specifically went wrong inside the function,
+but also what line of our code caused the original cascade. Here, the offending
+line is pointed to at the very top of the traceback, and by reading the error
+message at the bottom we can see that the problem was that we can't use the
+'less than' `<` comparison between two objects when one is a string
+(`dog_licenses['animal_name']`) and the other is a float (`dog_licenses['zip_code']`).
+
+## How can I operate on subsets of my data?
```

Comment:
What do you think about changing this heading to something like "How can I summarize and group my data?" or maybe "How can I summarize and operate on subsets of my data?"